File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/LongestConsecutiveSequence.py
#   Given [100, 4, 200, 1, 3, 2],
#   The longest consecutive elements sequence is [1, 2, 3, 4]. Return its length: 4.
#   Your algorithm should run in O(n) complexity.

class Solution:
    # @param num, a list of integer
    # @return an integer
    def longestConsecutive(self, num):
        """ find the longest consecutive sequence count """
        
        sequence_count = dict() 
        
        for idx in range(len(num)):
            sequence_count[num[idx]] = 1    # each number has sequence count 1
        
        longest_count = 1
        
        for idx in range(len(num)):
            if num[idx] in sequence_count:
                curr_value = num[idx] + 1
                while  curr_value  in sequence_count:
                    sequence_count[ num[idx] ] += sequence_count[ curr_value ]
                    del sequence_count[curr_value]  # delete curr_value for dup count
                    curr_value = curr_value + 1     # seach for ajacent bigger number

                longest_count = max( longest_count, sequence_count[ num[idx] ] )

        return longest_count

if __name__ == "__main__":
    test_nums = [-1,1,0]
    inst = Solution()
    print inst.longestConsecutive(test_nums)

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/CPP/LongestConsecutiveSequence.cpp
//Given an unsorted array of integers, find the length of the longest consecutive elements sequence.

//For example,
//Given [100, 4, 200, 1, 3, 2],
//The longest consecutive elements sequence is [1, 2, 3, 4]. Return its length: 4.
//Your algorithm should run in O(n) complexity.


class Solution {
public:
    int longestConsecutive(vector<int> &num) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        
        unordered_map<int, int> hashmap;
        for( int i=0; i<num.size(); i++)
            hashmap[ num[i] ] = 1;
            
        int ans=1;
        
        for( int i=0; i<num.size(); i++){
            int temp = num[i] + 1;
            auto it = hashmap.find( temp );
            while( it != hashmap.end() ){
                hashmap[ num[i] ] += it->second;
                hashmap.erase(it);
                it = hashmap.find( ++ temp );
            }
            
            ans = max( ans, hashmap[ num[i] ] );
        }
        return ans;    
    }
};
